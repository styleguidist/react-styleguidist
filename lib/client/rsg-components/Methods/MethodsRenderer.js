import "core-js/modules/es.symbol";
import "core-js/modules/es.symbol.description";
import "core-js/modules/es.function.name";
import "core-js/modules/es.object.assign";

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

import React from 'react';
import PropTypes from 'prop-types';
import Markdown from 'rsg-components/Markdown';
import Argument from 'rsg-components/Argument';
import Arguments from 'rsg-components/Arguments';
import Name from 'rsg-components/Name';
import JsDoc from 'rsg-components/JsDoc';
import Table from 'rsg-components/Table';

var getRowKey = function getRowKey(row) {
  return row.name;
};

export var columns = [{
  caption: 'Method name',
  // eslint-disable-next-line react/prop-types
  render: function render(_ref) {
    var name = _ref.name,
        _ref$tags = _ref.tags,
        tags = _ref$tags === void 0 ? {} : _ref$tags;
    return /*#__PURE__*/React.createElement(Name, {
      deprecated: !!tags.deprecated
    }, name + "()");
  }
}, {
  caption: 'Parameters',
  // eslint-disable-next-line react/prop-types
  render: function render(_ref2) {
    var _ref2$params = _ref2.params,
        params = _ref2$params === void 0 ? [] : _ref2$params;
    return /*#__PURE__*/React.createElement(Arguments, {
      args: params
    });
  }
}, {
  caption: 'Description',
  // eslint-disable-next-line react/prop-types
  render: function render(_ref3) {
    var description = _ref3.description,
        returns = _ref3.returns,
        _ref3$tags = _ref3.tags,
        tags = _ref3$tags === void 0 ? {} : _ref3$tags;
    return /*#__PURE__*/React.createElement("div", null, description && /*#__PURE__*/React.createElement(Markdown, {
      text: description
    }), returns && /*#__PURE__*/React.createElement(Argument, _extends({
      block: true,
      returns: true
    }, returns)), /*#__PURE__*/React.createElement(JsDoc, tags));
  }
}];

var MethodsRenderer = function MethodsRenderer(_ref4) {
  var methods = _ref4.methods;
  return /*#__PURE__*/React.createElement(Table, {
    columns: columns,
    rows: methods,
    getRowKey: getRowKey
  });
};

MethodsRenderer.propTypes = {
  methods: PropTypes.arrayOf(PropTypes.shape({
    name: PropTypes.string.isRequired,
    description: PropTypes.string,
    returns: PropTypes.object,
    params: PropTypes.array,
    tags: PropTypes.object
  }).isRequired).isRequired
};
export default MethodsRenderer;