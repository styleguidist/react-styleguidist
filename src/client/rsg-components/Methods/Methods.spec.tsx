import React from 'react';
import PropTypes from 'prop-types';
import { parse, MethodDescriptor } from 'react-docgen';
import { createRenderer } from 'react-test-renderer/shallow';
import MethodsRenderer, { columns } from './MethodsRenderer';

// Test renderers with clean readable snapshot diffs
export default function ColumnsRenderer({ methods }: { methods: MethodDescriptor[] }) {
	return (
		<ul>
			{methods.map((row, rowIdx) => (
				<li key={rowIdx}>
					{columns.map((col, colIdx) => (
						<div key={colIdx}>{col.render(row)}</div>
					))}
				</li>
			))}
		</ul>
	);
}

ColumnsRenderer.propTypes = {
	methods: PropTypes.array,
};

function render(methods: string[]) {
	const parsed = parse(
		`
		import { Component } from 'react';
		export default class Cmpnt extends Component {
			${methods.join('\n')}
			render() {
			}
		}
	`,
		undefined,
		undefined,
		{ filename: '' }
	);
	const renderer = createRenderer();
	if (Array.isArray(parsed) || !parsed.methods) {
		renderer.render(<div />);
	} else {
		renderer.render(<ColumnsRenderer methods={parsed.methods} />);
	}
	return renderer.getRenderOutput();
}

describe('MethodsRenderer', () => {
	it('should render a table', () => {
		const renderer = createRenderer();
		renderer.render(
			<MethodsRenderer
				methods={[
					{
						name: 'method',
						modifiers: [],
						params: [],
						description: 'Public',
					},
				]}
			/>
		);

		expect(renderer.getRenderOutput()).toMatchSnapshot();
	});
});

describe('PropsRenderer', () => {
	it('should render public method', () => {
		const actual = render(['/**\n * Public\n * @public\n */\nmethod() {}']);

		expect(actual).toMatchSnapshot();
	});

	it('should render parameters', () => {
		const actual = render([
			'/**\n * Public\n * @public\n * @param {Number} value - Description\n */\nmethod(value) {}',
		]);

		expect(actual).toMatchSnapshot();
	});

	it('should render returns', () => {
		const actual = render([
			'/**\n * @public\n * @returns {Number} - Description\n */\nmethod() {}',
		]);

		expect(actual).toMatchSnapshot();
	});

	it('should render JsDoc tags', () => {
		const renderer = createRenderer();
		renderer.render(
			<ColumnsRenderer
				methods={[
					{
						name: 'Foo',
						tags: {
							since: [
								{
									title: 'since',
									description: '1.0.0',
								},
							],
						},
					},
				]}
			/>
		);

		expect(renderer.getRenderOutput()).toMatchSnapshot();
	});

	it('should render deprecated JsDoc tags', () => {
		const renderer = createRenderer();
		renderer.render(
			<ColumnsRenderer
				methods={[
					{
						name: 'Foo',
						tags: {
							deprecated: [
								{
									title: 'description',
									description: 'Use *another* method',
								},
							],
						},
					},
				]}
			/>
		);

		expect(renderer.getRenderOutput()).toMatchSnapshot();
	});
});
