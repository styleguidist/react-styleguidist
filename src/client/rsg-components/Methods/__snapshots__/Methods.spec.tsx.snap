// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MethodsRenderer should render a table 1`] = `
<Styled(Table)
  columns={
    Array [
      Object {
        "caption": "Method name",
        "render": [Function],
      },
      Object {
        "caption": "Parameters",
        "render": [Function],
      },
      Object {
        "caption": "Description",
        "render": [Function],
      },
    ]
  }
  getRowKey={[Function]}
  rows={
    Array [
      Object {
        "description": "Public",
        "modifiers": Array [],
        "name": "method",
        "params": Array [],
      },
    ]
  }
/>
`;

exports[`PropsRenderer should render JsDoc tags 1`] = `
<ul>
  <li>
    <div>
      <Styled(Name)
        deprecated={false}
      >
        Foo()
      </Styled(Name)>
    </div>
    <div>
      <Styled(Arguments)
        args={Array []}
      />
    </div>
    <div>
      <div>
        <JsDoc
          since={
            Array [
              Object {
                "description": "1.0.0",
                "title": "since",
              },
            ]
          }
        />
      </div>
    </div>
  </li>
</ul>
`;

exports[`PropsRenderer should render deprecated JsDoc tags 1`] = `
<ul>
  <li>
    <div>
      <Styled(Name)
        deprecated={true}
      >
        Foo()
      </Styled(Name)>
    </div>
    <div>
      <Styled(Arguments)
        args={Array []}
      />
    </div>
    <div>
      <div>
        <JsDoc
          deprecated={
            Array [
              Object {
                "description": "Use *another* method",
                "title": "description",
              },
            ]
          }
        />
      </div>
    </div>
  </li>
</ul>
`;

exports[`PropsRenderer should render parameters 1`] = `
<ul>
  <li>
    <div>
      <Styled(Name)
        deprecated={false}
      >
        method()
      </Styled(Name)>
    </div>
    <div>
      <Styled(Arguments)
        args={
          Array [
            Object {
              "description": "Description",
              "name": "value",
              "optional": false,
              "type": Object {
                "name": "Number",
              },
            },
          ]
        }
      />
    </div>
    <div>
      <div>
        <Markdown
          text="Public"
        />
        <JsDoc />
      </div>
    </div>
  </li>
</ul>
`;

exports[`PropsRenderer should render public method 1`] = `
<ul>
  <li>
    <div>
      <Styled(Name)
        deprecated={false}
      >
        method()
      </Styled(Name)>
    </div>
    <div>
      <Styled(Arguments)
        args={Array []}
      />
    </div>
    <div>
      <div>
        <Markdown
          text="Public"
        />
        <JsDoc />
      </div>
    </div>
  </li>
</ul>
`;

exports[`PropsRenderer should render returns 1`] = `
<ul>
  <li>
    <div>
      <Styled(Name)
        deprecated={false}
      >
        method()
      </Styled(Name)>
    </div>
    <div>
      <Styled(Arguments)
        args={Array []}
      />
    </div>
    <div>
      <div>
        <Styled(Argument)
          block={true}
          description="Description"
          returns={true}
          type={
            Object {
              "name": "Number",
            }
          }
        />
        <JsDoc />
      </div>
    </div>
  </li>
</ul>
`;
